sasdata/dataloader/data_info.py

```diff
@@ -828,84 +836,111 @@ def copy_from_datainfo(self, data1d):
         """
-        Checks that the data lengths are compatible.
-        Checks that the x vectors are compatible.
-        Returns errors vectors equal to original
-        errors vectors if they were present or vectors
-        of zeros when none was found.
+        Checks that x values for two datasets have overlapping ranges for an operation.
+        If so, x, y, and dy will be updated to values used in the operation.
+        Resolutions, including dx, dxl, and dxw, are not kept through in the operation.
+        Wavelength parameters for SESANS datasets are also not kept through the operation.
 
-        :param other: other data set for operation
-        :return: dy for self, dy for other [numpy arrays]
-        :raise ValueError: when lengths are not compatible
+        :param other: other data for operation
+        :param tolerance: acceptable deviation in matching x data points, default 0.01 (equivalent to 1 % deviation)
+        :param scale: default 'log', performs interpolation on log scale; use 'linear' for sesans data
+        :raise ValueError: x-ranges of self and other do not overlap
         """
-        dy_other = None
         if isinstance(other, Data1D):
-            # Check that data lengths are the same
-            if len(self.x) != len(other.x) or len(self.y) != len(other.y):
-                msg = "Unable to perform operation: data length are not equal"
+            # check if ranges of self.x and other.x overlap at all
+            if np.min(other.x) > np.max(self.x) or np.max(other.x) < np.min(self.x):
+                msg = "Unable to perform operation: x ranges of two datasets do not overlap."
                 raise ValueError(msg)
-            # Here we could also extrapolate between data points
-            TOLERANCE = 0.01
-            for i in range(len(self.x)):
-                if fabs(self.x[i] - other.x[i]) > self.x[i]*TOLERANCE:
-                    msg = "Incompatible data sets: x-values do not match"
-                    raise ValueError(msg)
-
-            # Check that the other data set has errors, otherwise
-            # create zero vector
-            dy_other = other.dy
-            if other.dy is None or (len(other.dy) != len(other.y)):
-                dy_other = np.zeros(len(other.y))
-
-        # Check that we have errors, otherwise create zero vector
-        dy = self.dy
-        if self.dy is None or (len(self.dy) != len(self.y)):
-            dy = np.zeros(len(self.y))
+            # check if data points match (within tolerance) across the overlap range of self.x and other.x
+            self_overlap_bool = np.abs((self.x[:, None] - other.x[None, :]) / self.x[:, None]).min(axis=1) <= tolerance
+            self_overlap_index = np.flatnonzero(self_overlap_bool)
+            if len(self_overlap_index) == self_overlap_index.max() - self_overlap_index.min() + 1:
+                # all the points in overlap region of self.x found close matches in overlap region of other.x
```

Comment:
I don't really understand what exactly this is doing (I'm just a poor white belt) but it does seem that the goal here was to either do no interpolation if the two data sets have exactly the same number of points AND all the points match up in x within tolerance, otherwise ALL the data will be interpolated even if the x points match up? Given the funtionality review of the paired sasview PR is this really what is happening?